jib-core/src/main/java/com/google/cloud/tools/jib/registry/ManifestChecker.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.registry;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.api.client.http.HttpMethods;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.cloud.tools.jib.api.DescriptorDigest;
+import com.google.cloud.tools.jib.hash.Digests;
+import com.google.cloud.tools.jib.http.BlobHttpContent;
+import com.google.cloud.tools.jib.http.Response;
+import com.google.cloud.tools.jib.http.ResponseException;
+import com.google.cloud.tools.jib.image.json.ManifestTemplate;
+import com.google.cloud.tools.jib.image.json.OciManifestTemplate;
+import com.google.cloud.tools.jib.image.json.UnknownManifestFormatException;
+import com.google.cloud.tools.jib.image.json.V21ManifestTemplate;
+import com.google.cloud.tools.jib.image.json.V22ManifestListTemplate;
+import com.google.cloud.tools.jib.image.json.V22ManifestTemplate;
+import com.google.cloud.tools.jib.json.JsonTemplateMapper;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+/** Checks an image's manifest. */
+public class ManifestChecker<T extends ManifestTemplate>
+    implements RegistryEndpointProvider<Optional<ManifestAndDigest<T>>> {
+
+  private final RegistryEndpointRequestProperties registryEndpointRequestProperties;
+  private final String imageDescriptor;
+  private final Class<T> manifestTemplateClass;
+
+  ManifestChecker(
+      RegistryEndpointRequestProperties registryEndpointRequestProperties,
+      String imageDescriptor,
+      Class<T> manifestTemplateClass) {
+    this.registryEndpointRequestProperties = registryEndpointRequestProperties;
+    this.imageDescriptor = imageDescriptor;
+    this.manifestTemplateClass = manifestTemplateClass;
+  }
+
+  @Nullable
+  @Override
+  public BlobHttpContent getContent() {
+    return null;
+  }
+
+  @Override
+  public List<String> getAccept() {
```

Comment:
I don't think we should be duplicating code like this, the code in ManifestPuller and this can go out of sync. Is there a way provide a shared resource that both can refrence?

Reply:
Agreed. I really think we should not duplicate this much amount of code.

Reply:
I don't see an easy way how to extract common code from the 2 classes, given that they implement the `RegistryEndpointProvider` interface with 2 different types. A simple option would be to change the implementation of `RegistryClient.checkImage` to use `RegistryClient.pullManifest` and catch the exception, like so:

```
  public Optional<ManifestAndDigest<ManifestTemplate>> checkImage(String imageDigest) {
    try {
      return Optional.ofNullable(pullManifest(imageDigest, ManifestTemplate.class));
    } catch (RegistryErrorException e) {
      return Optional.empty();
    }
  }
```

What do you think?

Reply:
Would you mind if I push changes to your PR?

Reply:
Why would I? 😄

Reply:
Discussed this with @loosebazooka. I can take care of code duplication in a follow-up PR.

Reply:
I've just taken care of this in this PR.